connect/runtime/src/test/java/org/apache/kafka/connect/runtime/standalone/StandaloneHerderTest.java

```diff
@@ -123,91 +134,91 @@ private enum SourceSink {
         expectAdd(SourceSink.SOURCE);
 
         Map<String, String> config = connectorConfig(SourceSink.SOURCE);
-        Connector connectorMock = PowerMock.createMock(SourceConnector.class);
+        Connector connectorMock = mock(SourceConnector.class);
         expectConfigValidation(connectorMock, true, config);
 
-        PowerMock.replayAll();
-
         herder.putConnectorConfig(CONNECTOR_NAME, config, false, createCallback);
         Herder.Created<ConnectorInfo> connectorInfo = createCallback.get(1000L, TimeUnit.SECONDS);
         assertEquals(createdInfo(SourceSink.SOURCE), connectorInfo.result());
-
-        PowerMock.verifyAll();
     }
 
     @Test
     public void testCreateConnectorFailedValidation() {
         // Basic validation should be performed and return an error, but should still evaluate the connector's config
-        connector = PowerMock.createMock(BogusSourceConnector.class);
+        connector = mock(BogusSourceConnector.class);
 
         Map<String, String> config = connectorConfig(SourceSink.SOURCE);
         config.remove(ConnectorConfig.NAME_CONFIG);
 
-        Connector connectorMock = PowerMock.createMock(SourceConnector.class);
-        EasyMock.expect(worker.configTransformer()).andReturn(transformer).times(2);
-        final Capture<Map<String, String>> configCapture = EasyMock.newCapture();
-        EasyMock.expect(transformer.transform(EasyMock.capture(configCapture))).andAnswer(configCapture::getValue);
-        EasyMock.expect(worker.getPlugins()).andReturn(plugins).times(3);
-        EasyMock.expect(plugins.compareAndSwapLoaders(connectorMock)).andReturn(delegatingLoader);
-        EasyMock.expect(plugins.newConnector(EasyMock.anyString())).andReturn(connectorMock);
+        Connector connectorMock = mock(SourceConnector.class);
+        when(worker.configTransformer()).thenReturn(transformer);
+        final ArgumentCaptor<Map<String, String>> configCapture = ArgumentCaptor.forClass(Map.class);
+        when(transformer.transform(configCapture.capture())).thenAnswer(invocation -> configCapture.getValue());
+        when(worker.getPlugins()).thenReturn(plugins);
+        when(plugins.compareAndSwapLoaders(connectorMock)).thenReturn(delegatingLoader);
+        when(plugins.newConnector(anyString())).thenReturn(connectorMock);
 
-        EasyMock.expect(connectorMock.config()).andStubReturn(new ConfigDef());
+        when(connectorMock.config()).thenReturn(new ConfigDef());
 
         ConfigValue validatedValue = new ConfigValue("foo.bar");
-        EasyMock.expect(connectorMock.validate(config)).andReturn(new Config(singletonList(validatedValue)));
-        EasyMock.expect(Plugins.compareAndSwapLoaders(delegatingLoader)).andReturn(pluginLoader);
-
-        PowerMock.replayAll();
+        when(connectorMock.validate(config)).thenReturn(new Config(singletonList(validatedValue)));
+        when(Plugins.compareAndSwapLoaders(delegatingLoader)).thenReturn(pluginLoader);
```

Comment:
So interestingly if you remove this completely the tests also passes (including the other instances of `when(Plugins.compareAndSwapLoaders(delegatingLoader)).thenReturn(pluginLoader);`), so it appears that beforehand with EasyMock it wasn't even doing anything and with the migration to Mockito its also not doing anything.

Should I just remove all of the cases of `when(Plugins.compareAndSwapLoaders(delegatingLoader)).thenReturn(pluginLoader);` or investigate further why the mock in general is not using the static method?

Reply:
I did some research and it turns out that Mockito's static mocking API only supports setting expectations on the same thread that the `MockedStatic` was created on (see https://github.com/mockito/mockito/issues/2142).

I'm afraid to install a mocked classloader as the context classloader, even in tests, which is what's happening right now (you can verify this by adding a breakpoint in `Plugins::compareAndSwapLoaders` at the line that calls `Thread.currentThread()::setContextClassLoader` and running the tests through a debugger).

Can you see if there's a way to avoid installing a mocked classloader as the context classloader in these tests? Feel free to make small tweaks to non-testing code if necessary.

Reply:
Thanks for figuring this out, I will investigate further. One thing that is still confusing for me (which I might figure out later) is why the tests are still passing with the mocked static method not even being used? Ideally the tests at least currently should fail since the static mock isn't even being called, either that or mocking `DelegatingClassLoader` is "pointless".

Reply:
I added one more commit after `Create static mock on Herder's connectorExecutor and remove other unneeded mocks` called `Put back all the verifications back for compareAndSwapLoaders` which adds the verifications that I removed in the `Create static mock on Herder's connectorExecutor and remove other unneeded mocks` commit.

This commit now properly verifies that **all** of the mocks end up calling `Plugins.compareAndSwapLoaders` and `plugins.compareAndSwapLoaders`.

Reply:
Thanks @mdedetrich. I really appreciate the time you've taken to preserve the guarantees we get from these tests, without introducing other complications like switching the thread's context classloader to a mocked classloader during testing.

I'm still not sure this is safe to merge as-is, though. The `runOnHerderConnectorExecutor` method is fairly brittle, and may not serve our needs in the future if we end up needing to test concurrent connector validations. In addition, it does not prevent classloaders from being swapped out on other threads, such as the main testing thread or ones created and used by the standalone herder's `requestExecutorService`.

These issues are not impossible to work around; however, they would add a fair amount of clutter to our unit tests, and make it harder to read and write new tests in the future.

Instead, I've filed KAFKA-14346 with a proposal that we eliminate these static method-based APIs altogether. This isn't the first time that they've caused significant friction with our unit tests, and it's likely that it won't be the last. Instead of investing more effort in accommodating these APIs with changes to our testing logic, I think it's time to change the APIs to be easier to test.

Are you comfortable with either taking on the work of that ticket yourself, or pausing work on this PR until it has been completed?

Reply:
> I'm still not sure this is safe to merge as-is, though. The runOnHerderConnectorExecutor method is fairly brittle, and may not serve our needs in the future if we end up needing to test concurrent connector validations. In addition, it does not prevent classloaders from being swapped out on other threads, such as the main testing thread or ones created and used by the standalone herder's requestExecutorService.

Agreed, I do believe the reason why the test happens to work right now is that the current default `CachedThreadPool` happens to always use a single thread due to the fact that the tests and validations are run in serial. I am not sure of the exact details behind why Mockito's mockstatic only works for the thread where the mock was made, my initial assumption is that its due to using `ThreadLocal` which if is the case then the `CachedThreadPool` may have theoretically worked even for concurrent tests/validations due to them sharing the same `ThreadLocal` (which iirc is the default implementation behind `CachedThreadPool`).

I do definitely agree that its not a nice a solution, the user experience for people contributing tests is not ideal since they always have to remember to run static mocks on the `connectorExecutor` and as we have seen in the progress of the PR, if you forget to do this the tests can still inadvertently pass. 

I don't have the time right at this moment to work on such a feature but next week it may be better. If someone else wants to have a look at this then they should feel free to do so, don't have a problem with rebasing this PR once such a change lands.

Reply:
👍 It looks like @gharris1727 has self-assigned KAFKA-14346, so no need to take time off to work on that. Thanks for understanding, @mdedetrich; hoping we can move quicker and more confidently on this migration initiative once we've fixed the static mocking issue once and for all.

Reply:
@mdedetrich FYI, I've just merged a PR that addresses KAFKA-14346. This PR should be unblocked now; feel free to ping me after rebasing and I can start reviewing again.